doc/memos/rdm-draft-alamos-ieee802154-radio-hal.md

```diff
@@ -0,0 +1,908 @@
+The Radio HAL provides an Event Notification mechanism to inform the upper
+layer about an event (a packet was received, a transmission finished, etc).
+
+The upper layer can subscribe to these events to perform different actions. As
+an example, a MAC layer would subscribe to the RX done event to allocate the
+received packet. The TX done event is commonly used to release resources or
+update statistics.
+
+The full list of events and implications are defined in the Interface
+Definition section.
+
+### 4.3.3 Device Specific IEEE802.15.4 HAL Implementation
+The Device Specific IEEE802.15.4 HAL implementation is part of the IEEE802.15.4
+Radio HAL component in the above figure. It implements the hardware-dependent
+part of the IEEE802.15.4 Radio HAL by wrapping the `radio_ops` interface around
+the device specific code by using the Device Driver API which grants access to
+all device operations.
+
+## 4.4 Device Driver
+The Device Driver implements the Hardware Adoption Layer of the device. It
+provides minimal functionalities needed by the Radio Operations,
+and additionally a mechanism to expose the ISR of the radio, so the
+Bottom-Half processor can offload the ISR.
+
+The function set of the Device Driver can include device specific features that
+are not exposed but the Radio HAL API (e.g., Smart Listening with AT86RF2xx
+radios).  The Device Driver is an independent component and it can be used
+without the Radio HAL on top, for testing purposes or device specific
+applications.
+
+# 5 Implementation Details
+## 5.1 Initialization and Start of Device Drivers
+In order to implement the 802.15.4 abstraction on top of a device driver, it
+is required to separate the initialization process of the device in two
+stages:
+- Init: This stage sets up the device driver (analogue to the current
+  `xxx_setup` functions) and puts it in a state that minimizes power
+  consumption.
+- Start: This stage is intended to be triggered during network interface
+  initialization (e.g. via `ifconfig up`) to put the device in a state where it
+  is ready to operate (enable IRQ lines, turn on the transceiver, etc).
+
+Explicitly separating the Init and Start process is more efficient in terms of
+power consumption, because a network stack might take some time to set an
+interface up. The `radio_ops` interface provides a "start" function that should
+be mapped to the devices Start function proposed above.
+
+## 5.2 Transceiver States
+Following the IEEE802.15.4 PHY definition, there are three generic transceiver
+states:
+
+```c
+typedef enum {
+    IEEE802154_TRX_STATE_TRX_OFF, /**< the transceiver state if off */
+    IEEE802154_TRX_STATE_RX_ON, /**< the transceiver is ready to receive */
+    IEEE802154_TRX_STATE_TX_ON, /**< the transceiver is ready to send  */
+} ieee802154_trx_state_t;
+```
+
+Unlike the `netdev` interface, the 802.15.4 Radio HAL requires only these three
+states to drive the radio. This is due to the fact that the radio is either in
```

Comment:
Not true. The programmer needs to take into account that a frame can only be fetched after it has been received. So the actual number of states to take into account is larger. (See above.)

Reply:
I think what I wrote here is not accurate (see https://github.com/RIOT-OS/RIOT/pull/13943#discussion_r415345585)

To me, this is an upper layer logic. The HAL guarantees that it's safe to access the frame buffer on RX_DONE. But the transceiver state could be TRX_OFF or TX_ON, since the FB access is independent of the state.

That's what I mean with the fact this are only "transceiver states" (PHY states). In fact, these states are mostly independent to MAC and device states (the exception is the TRX_OFF during sleep, since it's implicit that the transceiver is disabled when the device sleeps)

Reply:
You're correct. Only three states are sufficient, when only looking at the transceiver. (But you might want to highlight a bit more that these states should only cover the transceiver?)

Reply:
Sure, I will clarify that